cayenne/src/main/java/org/apache/cayenne/access/translator/select/DefaultSelectTranslator.java

```diff
@@ -40,11 +40,11 @@ public class DefaultSelectTranslator implements SelectTranslator {
     private static final TranslationStage[] TRANSLATION_STAGES = {
             new ColumnExtractorStage(),
             new PrefetchNodeStage(),
-            new OrderingStage(),
             new QualifierTranslationStage(),
             new HavingTranslationStage(),
-            new GroupByStage(),
             new DistinctStage(),
+            new OrderingStage(),    // Relies on DistinctStage
```

Comment:
This change is not safe, logic in the `DistinctStage` relies on joins defined, while ordering expression translation could add mode joins. Probably we should try to split `OrderingStage` into two parts.

Reply:
Another possible solution is to move the bulk of the logic in `DistinctStage.perform` to `TranslatorContext`, then I can revert this change. What do you think ?

Reply:
Actually it may be better moving it into a separate `DistinctUtil` class ?

Reply:
Or combine this with `hasAggregate` from GroupByStage into a `StageUtil` class ?

Reply:
I've restored the original Stage ordering in DefaultSelectTranslator and instead moved all the shared logic code into a StageUtil class. This process revealed that Orderings with compound paths now had to be explicitly dealt with in OrderingStage :-)